skbio/io/format/biom.py

```diff
@@ -28,13 +30,19 @@
 
 Examples
 --------
-Here we will write an existing BIOM table, and re-read it. Note that the Table
-from ``biom`` implicitly gets the ``.write`` method from the IO registry.
-
->>> import biom, skbio
->>> biom.example_table.write('test.biom')
-'test.biom'
->>> roundtrip = skbio.read('test.biom', into=skbio.Table)
+Here we're demonstrating how to convert a BIOM table into HDF5 format and then
+read it back.
+
+>>> import io
+>>> import biom
+>>> import skbio
+>>> import h5py
+>>> f = io.BytesIO()
+>>> with h5py.File(f, 'w') as fp:
+...     biom.example_table.to_hdf5(fp, 'asd')
+>>> roundtrip = skbio.read(io.BytesIO(f.getvalue()), into=skbio.Table)
```

Comment:
What if `io.BytesIO(f.getvalue())` is replaced with `f`? If not working, what if adding `f.seek(0)` beforehand?

Reply:
I have this written now:

````
>>> import io
>>> import skbio
>>> f = io.BytesIO()
>>> skbio.feature_table.example_table.write(f)
<_io.BytesIO object at 0x7f969d056390>
>>> f.seek(0)
>>> roundtrip = skbio.read(f, into=skbio.Table)
>>> roundtrip
2 x 3 <class 'biom.table.Table'> with 5 nonzero entries (83% dense)
````

But the memory address of the `BytesIO` object is different when I run `make test` compared to when I just run it in python, which makes the test fail. I'm not very experienced with this so I'm not sure how to proceed.

Reply:
what about [`skip`](https://docs.pytest.org/en/7.1.x/how-to/doctest.html#skipping-tests)?

Reply:
Importantly, all of the following only occurs when I run `make test`. When I perform the same commands in the Python shell the errors don't occur and `f` correctly contains the content of `skbio.feature_table.example_table`.

````
>>> skbio.feature_table.example_table.write(f)  # doctest: +SKIP
<_io.BytesIO object at 0x7f969d056390>
>>> print(f.getvalue())
Expected nothing
Got:
    b''
````

It appears that `f` is an `<emptyfile>` while it should contain the content of `skbio.feature_table.example_table`, if my understanding is correct. This leads to the following error.

````
>>> roundtrip = skbio.read(f, into=skbio.Table)
UNEXPECTED EXCEPTION: UnrecognizedFormatError("Cannot read '<emptyfile>' from <WrappedBufferedRandom>, no Table reader found.  ")
````

This error occurs whether I have `f.seek(0)` preceding the `roundtrip = skbio.read(f, into=skbio.Table)` line or not. 

Thank you for the help so far @wasade.

Reply:
Doesn't look empty? I'm not sure what `getvalue` does off hand, but a `seek(0)` followed by `read` confirms there are quite a few bytes of data, and I'm able to load from a `seek(0)`:

```
>>> import skbio
>>> import io
>>> f = io.BytesIO()
>>> skbio.feature_table.example_table.write(f)
<_io.BytesIO object at 0x7fa6a014e890>
>>> f.seek(0)
0
>>> what = f.read()
>>> len(what)
37914
>>> f.seek(0)
0
>>> tab = skbio.read(f, into=skbio.feature_table.Table)
>>>
```

Reply:
I agree, when running it interactively it works and I get the same result as you. However, I am still encountering the above errors when I run the test by the command `make test`:

````
========================================== FAILURES ==========================================
_______________________________ [doctest] skbio.io.format.biom _______________________________
035
036 >>> import io
037 >>> import skbio
038 >>> f = io.BytesIO()
039 >>> skbio.feature_table.example_table.write(f)  # doctest: +SKIP
040 <_io.BytesIO object at 0x7ff6309e0f40>
041 >>> f.seek(0)
042 0
043 >>> what = f.read()
044 >>> len(what)
Expected:
    37914
Got:
    0

/home/maton/qlab/skbio/github/scikit-bio/skbio/io/format/biom.py:44: DocTestFailure
````

Reply:
The most recent commit works. The issue was when running `make test` I was skipping the line which wrote the file by using `doctest: +SKIP`. Instead, I used `doctest: +ELLIPSIS` and this works. Thank you again @wasade!

Reply:
Thanks for all the hard work!! I know very much so how tedious doctest can get :/